### PR TITLE
Refactor stock exit for performance and clarity

### LINKED ISSUE #5577: Refactor Stock Exit controller
Our stock exit controller has grown increasingly complex as we've added date checks, expiry checks, quantity checks...  We should take a look at refactoring it for clarity and efficiency into something we can unit test, similar to how [Invoices](https://github.com/IMA-WorldHealth/bhima/blob/6ad5804ba9c67b3a2b12335fae2b1a922c9cfcd2/test/client-unit/services/PatientInvoiceForm.spec.js) are tested.  It will also make adding new features (such as https://github.com/IMA-WorldHealth/bhima/issues/5576) a lot easier.

### PR DESCRIPTION
**Pull request by jniles**
The following major improvements are coming in this PR:

1. A rewritten `bhStockExitType` component that is smaller, cleaner, and has fewer bindings.  The component is now responsible for rendering its own label, rather than having multiple paths through the stock exit controller to render the label.  In the long run, we should have separate `bhStockExitType` and `bhStockEntryType` components.
2. The form logic is moved into an AngularJS service `StockExitFormService`.  This allows us to test things like the validation logic.
3. A new `LotItemService` is implemented for each grid row in the StockExit grid.  This allows us to handle the per-lot validation and the form validation a bit more cleanly.
4. Only a single call is made to the server for the stock lots!  Further, this route is rewritten to only JOIN on tables that we need and do the JOINs in a more optimal fashion.
5. I've written tests for the validation logic in the `LotItemService`.

### TODO
- [x] Finish the logic for removing items + revalidating the grid.
- [x] Finish and document the validation logic for the grid.  
    - [x] Add "actions" log under description showing the outstanding actions needed to be taken by a user to complete validation.
- [ ] Figure out how requisitions are fulfilled - right now it looks like the client is responsible for validating if a requisition has been fulfilled and that isn't probably correct.  Perhaps rewrite this portion of the code and move it server-side.  (See https://github.com/IMA-WorldHealth/bhima/issues/6424)
- [x] Implement `getLotsFromInventory()` feature to bring in lots from:
   - [x] Patient Invoice
   - [x] Service (requisition)
   - [x] Depot (requisition)
- [x] Submission
    - [x] Submit from Patient
    - [x] Submit from Loss
    - [x] Submit from Depot/Transfer
    - [x] Submit from Service
- [x] Rewrite description rendering logic - potentially in a `StockExitFormHelperService` to reduce file sizes.
- [x] Re-implement barcode scanning feature.
    - [x] Bring up the barcode scanning modal for the items
    - [x] Only choose from items in stock in the current depot.
    - [x] Display a prominent warning if the items are in stock in another depot. 
- [x] Write tests for LotItemService validation
- [ ] Write tests for StockExitFormService
- [X] Re-implement export to excel from grid

Closes #5577.

### CODE REVIEW

**Comment by jniles** [comment]
@jmcameron @mbayopanda I'm at the point where I need some reviews.

I have two major portions of code to work on:
  1. The barcode scanning feature (currently commented out)
  2. Importing from requisitions (depot, service).
  
There are a few tests I need to write for exporting the grid to Excel and such.  Once the logic all works, I'll work on better translations.  The French translations are non-existant, but the English could also be improved.
  
The rest should work.

Please try to break this code.  The sooner we can find all the bugs the better.  Please include screenshots when you find a bug.

File: client/src/js/components/bhStockExpired/bhStockExpired.js [hunk 1]
```diff
@@ -9,33 +9,40 @@ angular.module('bhima.components')
...
 
   $ctrl.$onInit = () => {
-    fetchExpiredStock();
-    getDepot();
+    $ctrl.loading = true;
+    $ctrl.expiredInventories = [];
+
+    $q.all([
+      fetchExpiredStock(),
+      getDepot(),
```
Comment (jmcameron) [inline client/src/js/components/bhStockExpired/bhStockExpired.js:28]: 👍 Good to put this into the promises

File: client/src/js/components/bhStockOut/bhStockOut.js [hunk 2]
```diff
@@ -49,8 +48,8 @@ function bhStockOutController(Depots, moment, Notify, $filter, $q) {
       .then(([inventories, depot]) => {
 
         inventories.forEach(inventory => {
-          inventory.stock_out_date_raw = $date(inventory.stock_out_date);
-          inventory.stock_out_date_parsed = moment(inventory.stock_out_date).fromNow();
+          inventory.stock_out_date_raw = $date(inventory.date);
```
Comment (jmcameron) [inline client/src/js/components/bhStockOut/bhStockOut.js:54]: Why change inventory.stock_out_date to inventory.date?

File: client/src/modules/stock/LotItem.service.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+    const clone = { ...item };
+    delete clone.uuid;
+
+    if (clone.tracking_consumption !== undefined) {
+      this.setTrackingConsumption(clone.tracking_consumption);
+      delete clone.tracking_consumption;
+    }
+
+    if (clone.tracking_expiration !== undefined) {
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js:268]: This pattern seems repetitive.  Is there any way to streamline it?

File: client/src/modules/stock/LotItem.service.js [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    toggleErrorMessage(!hasInfo, this._errors, ER_MISSING_INVENTORY);
+
+    return hasInfo;
+  };
+
+  /**
+   * @function hasLotInformation
+   *
+   * @description
+   * Checks if the lot information is availability.
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js]: Typo? (last word on line)

File: client/src/modules/stock/LotItem.service.js [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+   *
+   * 1) If the expiration date of the inventory should not be tracked (e.g. an asset)
+   *    we always return FALSE: the lot is not expired.  Note this condition applies
+   *    even if the lot has an expiration date and even if that expiration date is past.
+   * 2) If is_expired is set, we return that value rather than performing a
+   *    calculation.  This allows the server to have the same logic as the client and
+   *    override the logic here without overriding the error handling.
+   * 3) If a date comparison is passed in, the lot's expiration state is computed by
+   *    comparing to the comparison date.  If the lot's expiration_date  is before the
+   *    comparison date, it is considered expired.  If the lot's expiration_date is after,
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js:174]: what about "the same as" case?

File: client/src/modules/stock/LotItem.service.js [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    // the `is_expired` flag is set on the server.  If this
+    // flag is set, return the results of this flag.
+    } else if (this.is_expired !== undefined) {
+      isLotExpired = !!this.is_expired;
+
+    // if a date has been passed in, it will validate against that date
+    // (for example, if this is a past date).  If not, it will validate
+    // against the current date.
+    } else {
+      isLotExpired = this.expiration_date < comparisonDate;
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js:198]: 👍

File: client/src/modules/stock/LotItem.service.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+  /**
+   * @function isUnused
+   *
+   * @description
+   * Returns true if the lot has quantity available and is unused. This is to catch
+   * error cases where a user did not complete the form correctly.  However, if a lot
+   * is empty by default (no quantity in stock), no error is triggered.
+   */
+  Lot.prototype.isUnused = function isUnused() {
+    const hasEmptyQuantityAndUnused = this.isEmpty() && this._quantity_available !== 0;
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js:226]: Can we ever have a case where _quantity_available is negative?   If so, then we should use: > 0?

File: client/src/modules/stock/LotItem.service.js [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+  };
+
+  /**
+   * @function isConsumable
+   *
+   * @description
+   * Returns true if the lot is consumable.
+   */
+  Lot.prototype.isConsumable = function isConsumable() {
+    return this.consumable === 1;
```
Comment (jmcameron) [inline client/src/modules/stock/LotItem.service.js:238]: Or  use the !! syntax?

File: client/src/modules/stock/StockModal.service.js [hunk 9]
```diff
@@ -40,7 +41,6 @@ function StockModalService(Modal) {
     const params = angular.extend(modalParameters, {
       templateUrl  : 'modules/stock/entry/modals/generateAssetTags.modal.html',
       controller   : 'GenerateTagsModalController',
-      controllerAs : '$ctrl',
```
Comment (jmcameron) [inline client/src/modules/stock/StockModal.service.js:43]: Why were these removed?  Is there a default?

**Review by jmcameron — COMMENTED** [review]
I reviewed the code and left inline comments.

I tested it.  For the most part, the new Stock Exit works very nicely and is friendlier to the user.  I like the "Pending actions" section.

A few issues:
- If I add an inventory item, it selects the first available lot and inserts its quantity correctly.  However, if you at that point decide you want a different lot, when you select it, the quantity is zero (it is not updated).
- I tried doing an exit to Depot and selecting the inventory item worked fine, but the client side had errors when I tried to submit: TypeError: Cannot read properties of undefined (reading 'data')
- If I have started with "Loss", added an item, and decide I should do exit to "Depot", the inventory item grid should be erased (currently it is not).
- When I tried an exit to Service, and added an inventory item, it picked a lot that was expired.  I saw similar behavior for exit to Patent.

File: client/src/js/components/bhStockOut/bhStockOut.js (hunk 2 above)
Comment (jniles) [reply client/src/js/components/bhStockOut/bhStockOut.js:54]: To address a bug - the server sends back the `date` now, not the `stock_out_date`.  This is a bug in master, and I'm fixing it here, since I had to investigate it to see if it was me that broke it.

File: client/src/modules/stock/LotItem.service.js (hunk 7 above)
Comment (jniles) [reply client/src/modules/stock/LotItem.service.js:226]: Yes, we can, but by accident.  In several databases we have observed where the quantity in stock is negative, but it isn't supposed to happen.

File: client/src/modules/stock/StockModal.service.js (hunk 9 above)
Comment (jniles) [reply client/src/modules/stock/StockModal.service.js:43]: Yes, I added it to the default `modalParameters`.

File: client/src/modules/stock/LotItem.service.js (hunk 3 above)
Comment (jniles) [reply client/src/modules/stock/LotItem.service.js:268]: I can't think of a shorter version... can you?

**Comment by jniles** [comment]
@jmcameron @mbayopanda this is ready for a second round of reviews.  Try to break this module again. 

I've re-implemented the barcode scanning.

If it all behaves as you expect, then I'll need @mbayopanda's help on the translations for these error messages.

I've gotten the import from requisitions working, but not marking the requisitions are finalized. See #6424 for commentary on that.  If you guys are okay, I would propose we merge this as-is and then I'll work on refactoring the server-side code to make it more performant and cleaner, while also closing #6424.  Right now, this takes a very long time to create a stock movement on my machine.

**Review by mbayopanda — CHANGES_REQUESTED** [review]
This is big refactoring and the code is very clear now, I just noticed that with the test database when we test exit on the "Quinine Bichlorhydrate, sirop, 100mg base/5ml, 100ml, flacon, Unité", we have these issues:

- The available quantity is set to zero (0) when the lot is changed
- We can submit the stock exit with zero as quantity

**Comment by jniles** [comment]
@mbayopanda thank you for your testing and review!  I'll fix those issues ASAP.

**Comment by jniles** [comment]
@mbayopanda can you give more steps to reproduce the submitting a 0 quantity lot?

I have fixed the other issue.

**Comment by jniles** [comment]
@mbayopanda @jmcameron I think this is essentially ready to go.  I've also sped up the queries for creating movements on the server.  It can still be improved though.  Please take a look at this and give me your thoughts. 

If you are happy with it, we can merge it and I'll keep working on refactoring the server-side.

**Comment by mbayopanda** [comment]
## 1. Lot disappear in the list

Assume that we want to exit to loss expired lots, with the test database : 
- We have taken this Quinine A 
image

- In the exit page, the lot is loaded after selecting the inventory:
image

But after changing the lot to something like Quinine B, we cannot see Quinine A anymore
image

The behavior seems not normal, I don't know if you can reproduce that.

## 2. We can select the same lot multiple times

image

The quantity available for Quinine B was 15, but I can exit 16 like this: 
image

And here the negative value implication:
image

**Comment by mbayopanda** [comment]
@jniles I made some comments

File: client/src/modules/stock/LotItem.service.js (hunk 7 above)
Comment (jmcameron) [reply client/src/modules/stock/LotItem.service.js:226]: I realize that this is not supposed to happen.  But shouldn't we prevent that type of error from affecting this functionality by using '> 0' instead of '!== 0'?

File: client/src/js/services/util.js [hunk 10]
```diff
@@ -27,6 +27,11 @@ function UtilService(moment) {
     return null;
   };
 
+  service.getUniqueBy = function getUniqueBy(array, prop) {
```
Comment (jmcameron) [inline client/src/js/services/util.js:37]: Please add a descriptive comment for this function

**Review by jmcameron — COMMENTED** [review]
I took a brief glance at the code again.

I tested Stock Exit with bhima_test
- In Exit to stock to patient
   - once I selected a patient and an invoice, it (correctly) complained that
     one item was out of stock and would be skipped.  Excellent message!
   - If the invoice calls for an expired stock, it is still listed and the
     warning message is not very clear.  A warning (in the Pending Actions
     section) could say: This stock is expired, select another lot.  (or
     something similar).
   - Good message when an invoice could only partially be filled.
- When I tried Exit to Service, I selected a stock, etc, but clicking on
  the [Submit] button did nothing.
- Exit to Depot worked fine
- Exit to Loss worked fine

**Comment by jniles** [comment]
@jmcameron @mbayopanda, thanks for your reviews.  I've fixed all the bugs you guys pointed out.

I've done another pass to test all the common routes and they work.  Please check again for me as well.

### Performance Improvements

This PR contains significant performance improvements.  When the stock exit module is loaded on `master`, 5 requests are sent for the data.  This PR reduces it to 4.  Additionally, the longest request takes nearly 2 seconds (on my machine, HEV's database).  The longest query now takes ~250ms.  The raw data is shown below:

_master_

<pre> <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24 HTTP/1.1 304 - - 12.893 ms <font color="#5F5FD7">+93ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24/flags/stock_out?date=2022-03-03T18:37:36.038Z HTTP/1.1 200 11133 - 108.047 ms <font color="#5F5FD7">+95ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24/flags/expired?date=2022-03-03T18:37:36.038Z HTTP/1.1 200 1130 - 647.947 ms <font color="#5F5FD7">+541ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /stock/lots/depots?dateTo=2022-03-03T18:37:36.181Z&amp;depot_uuid=B2A92E50927548199925C43346A89E24&amp;skipTags=true HTTP/1.1 200 2449453 - 1400.773 ms <font color="#5F5FD7">+745ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /stock/inventories/depots?dateTo=2022-03-03T18:37:36.180Z&amp;depot_uuid=B2A92E50927548199925C43346A89E24&amp;is_expired=0&amp;skipTags=true HTTP/1.1 200 438566 - 1921.228 ms <font color="#5F5FD7">+490ms</font></pre>

_this branch_

<pre>  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24 HTTP/1.1 304 - - 10.648 ms <font color="#5F5FD7">+113ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24/flags/stock_out?date=2022-03-03 HTTP/1.1 304 - - 51.432 ms <font color="#5F5FD7">+40ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24/stock?consumable=1&amp;date=2022-03-03T18:34:37.338Z&amp;inStock=true HTTP/1.1 200 78660 - 89.668 ms <font color="#5F5FD7">+9ms</font>
  <font color="#5F5FD7"><b>http </b></font>::ffff:192.168.133.1 - GET /depots/B2A92E50927548199925C43346A89E24/flags/expired?date=2022-03-03 HTTP/1.1 304 - - 241.720 ms <font color="#5F5FD7">+187ms</font>
</pre>

I am optimistic this will greatly help the processing at Vanga and reduce server crashes.

**Review by mbayopanda — APPROVED** [review]
Good job

LGTM 👍🏽
